File: apps/atlas-api/src/features/plan-notices/application/get-plan-notices.ts
import type { BenchUser } from "../../identity/domain/user";
import { userIndex } from "../../identity/domain/user";

export type PlanNotice = {
  id: string;
  severity: "info" | "warn";
  message: string;
};

export type PlanNoticesResponse = {
  notices: PlanNotice[];
  plan: string;
};

const LIMITS: Record<BenchUser["plan"], number> = {
  free: 1_000,
  pro: 50_000,
  enterprise: 500_000,
};

/** Deterministic bench notices derived from the signed-in user. */
export function getPlanNotices(user: BenchUser): PlanNoticesResponse {
  const idx = userIndex(user.id);
  const requests = 120 + idx * 17;
  const limit = LIMITS[user.plan];
  const pct = Math.round((requests / limit) * 100);
  const notices: PlanNotice[] = [];

  if (pct >= 80) {
    notices.push({
      id: "limit-warning",
      severity: "warn",
      message: `You've used ${pct}% of your monthly request limit (${requests.toLocaleString()} / ${limit.toLocaleString()}).`,
    });
  }

  if (user.plan === "free") {
    notices.push({
      id: "upgrade-hint",
      severity: "info",
      message: "Upgrade to Pro for higher limits and priority support.",
    });
  }

  if (notices.length === 0) {
    notices.push({
      id: "all-clear",
      severity: "info",
      message: "Your plan is in good standing this month.",
    });
  }

  return { notices, plan: user.plan };
}

File: apps/lantern-api/src/features/plan_notices/application/get_plan_notices.py
from src.features.identity.domain.user import BenchUser, user_index

LIMITS = {
    "free": 1_000,
    "pro": 50_000,
    "enterprise": 500_000,
}


def get_plan_notices(user: BenchUser) -> dict:
    idx = user_index(user.id)
    requests = 120 + idx * 17
    limit = LIMITS[user.plan]
    pct = round((requests / limit) * 100)
    notices: list[dict] = []

    if pct >= 80:
        notices.append(
            {
                "id": "limit-warning",
                "severity": "warn",
                "message": (
                    f"You've used {pct}% of your monthly request limit "
                    f"({requests:,} / {limit:,})."
                ),
            }
        )

    if user.plan == "free":
        notices.append(
            {
                "id": "upgrade-hint",
                "severity": "info",
                "message": "Upgrade to Pro for higher limits and priority support.",
            }
        )

    if not notices:
        notices.append(
            {
                "id": "all-clear",
                "severity": "info",
                "message": "Your plan is in good standing this month.",
            }
        )

    return {"notices": notices, "plan": user.plan}
